[howenc] iP

howenc opened the pull request:
# **Duke Edits for Howen's IP**
## Description of changes and what they are supposed to do
- The program is built and tested using **gradle**
- I tried to do as much OOP as I can
- I have stud test classes that are just copies of the previous versions of my code that does not have any bugs

**Description of the classes I have**

1. Commands - the state of this class determines the executed task. So the instance of Commands that is created will already possess all that it needs to execute the intended task. I only configured it to take up to 3 phase inputs, phases being strings separated by "/". This way once I put `comandsInstance.execute()` it will do the task and return a Boolean to check if it succeed.

2. Parser - the Parser class is the class that turns the string input by the user and outputs a Commands object. How it works is that each phase of a input (string separated by "/") would be evaluated if needed, there would not be a NullPointerException thrown if the phase is not there. After getting the command words according to the enums COMMAND, it will do a check on the input and create and return the Commands instance accordingly.
```Java
Parser p = new Parser(String input);
Commands c = p.parse();
c.execute();
```

3. thats all the importing things. So much for
> **Description of the classes I have**
<!-- ![Screenshot of a comment on a GitHub issue showing an image, added in the Markdown, of an Octocat smiling and raising a tentacle.](https://myoctocat.com/assets/images/base-octocat.svg) -->

4. Click [here](https://www.youtube.com/watch?v=dQw4w9WgXcQ) ~~if you want to see more~~

## Logic behind design of certain classes
🤡 


- [x] a heading
- [x] a bullet list
- [x] a numbered list
- [x] a fenced code block (with syntax highlighting)
- [x] a task list
- [x] an emoji
- [x] a blockquote
- [x] a hyper link
- [x] inline code
- [x] some text formatting: bold, italic, stikethrough etc.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

howenc committed:
Added greeting and exit.

howenc committed:
Added nextCommand and echo method

howenc committed:
Added ability to store whatever text entered

howenc committed:
Added mark and unmark

howenc committed:
Added Todo, Deadline, Event

howenc committed:
Trimmed unnecessary code

howenc committed:
Added Parse class and edit nextCommand, exceptions.

howenc committed:
Implemented ArrayList<Task>

howenc committed:
Added delete

howenc committed:
Added enums for the commands

howenc committed:
Edited UI testing

howenc committed:
Removed extra indent to fit code style

howenc committed:
Fixed a mistake, added packaged main.java.Duke

howenc committed:
Changed public methods to protected

howenc committed:
Changed the package added a Task file

howenc committed:
Add quickLoad method and data.duke.txt file

howenc committed:
Add readData

howenc committed:
Minor print change

howenc committed:
Add writeData

howenc committed:
Handle corrputed data

howenc committed:
Merge branch 'branch-Level-7'

howenc committed:
Add LocalDateTime for tasks

howenc committed:
Minor changes to commands

howenc committed:
Add Parser, Storage, Ui, Commands classes

howenc committed:
Use state of Command Object as decision maker

howenc committed:
Minor edit

howenc committed:
Add twoCommands and threeCommands

howenc committed:
Save point 1

howenc committed:
Refactor code for OOP

howenc committed:
Merge branch 'A-MoreOOP'

howenc committed:
Merge branch 'branch-Level-8'

howenc committed:
Made packages

howenc committed:
Merge remote-tracking branch 'origin/add-gradle-support'

howenc committed:
Add Gradle

howenc committed:
Unmark main as module

howenc committed:
Gradle and JUnit setup

howenc committed:
Exceptions

howenc committed:
Test cases for all commands

howenc committed:
Made a Jar file

howenc committed:
Minor edit

howenc committed:
Minor edit

howenc committed:
Minor edit

howenc committed:
Minor edit

howenc committed:
JavaDoc and refactor for Commands.java

howenc committed:
Merge branch 'master' into branch-A-JavaDoc

howenc committed:
Writing JavaDocs

howenc committed:
Editing to match coding standard

howenc committed:
Made find command

howenc committed:
Merge branch 'branch-A-JavaDoc'

howenc committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/command/Commands.java
#	src/main/java/duke/Duke.java
#	src/main/java/task/ListOfTask.java
#	src/main/java/task/Task.java

howenc committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java

howenc committed:
Minor edits and checking

howenc committed:
Minor Edit

LoMaply commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,179 @@
+     * @param taskName The name of the task.
+     * @return Returns a Task object.
+     */
+    public static Task of(String taskName) {
+        return new ToDos(taskName);
+    }
+
+    /**
+     * Calls the constructor to create a Task.
+     *
+     * @param taskName The name of the task.
+     * @param dayDate The date and time of the deadline of the task.
+     * @return Returns a Task object.
+     */
+    public static Task of(String taskName, LocalDateTime dayDate) {
+        return new Deadlines(taskName, dayDate);
+    }
+
+    /**
+     * Calls the constructor to create a Task.
+     *
+     * @param taskName The name of the task.
+     * @param startDayDateTime The date and time of the start of the event.
+     * @param endDayDateTime The date and time of the end of the event.
+     * @return Returns a Task object.
+     */
+    public static Task of(String taskName, LocalDateTime startDayDateTime, LocalDateTime endDayDateTime) {
+        return new Event(taskName, startDayDateTime, endDayDateTime);
+    }
```
These constructors seem a little redundant, you could call the child class constructors directly after putting each child class into it's own file.

SelwynAng commented on src/main/java/command/Commands.java:
```diff
@@ -0,0 +1,351 @@
+    /**
+     * List of both primary and secondary commands.
+     */
+    public enum COMMANDS {BYE, LIST, TODO, DEADLINE, EVENT, MARK, UNMARK, DELETE, BY, FROM, TO, SORT, FIND, UNKNOWN}
```
I like how you classified the commands using enums too. Good job for going the extra mile!

LoMaply commented on src/main/java/task/ListOfTask.java:
```diff
@@ -0,0 +1,171 @@
+     * @param print True to print messages, false to not print messages.
+     * @throws DukeException If the number is outside the range of indexes in the list.
+     */
+    public void mark(int index, boolean print) throws DukeException {
+        try {
+            listOfTask.get(index - 1).setDone();
+            if (print) {
+                System.out.println(listOfTask.get(index - 1).toString());
+                Storage.save(listOfTask);
+            }
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeException("Please select from index 1 to " + listOfTask.size());
+        }
+    }
+
+    /**
+     * Mark a task as undone.
+     *
+     * @param index The index of the task based on the current task list.
+     * @param print True to print messages, false to not print messages.
+     * @throws DukeException If the number is outside the range of indexes in the list.
+     */
+    public void unMark(int index, boolean print) throws DukeException {
+        try {
+            listOfTask.get(index - 1).setNotDone();
+            if (print) {
+                System.out.println(listOfTask.get(index - 1).toString());
+            }
+            Storage.save(listOfTask);
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeException("Please select from index 1 to " + listOfTask.size());
+        }
+    }
```
To reduce some code duplication, you might consider merging both the `mark()` and `unMark()` methods and using an additional variable to decide whether to mark or unmark a particular task.

howenc replied:
Got it

LoMaply commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,100 @@
+            writeData.createNewFile();
+            FileWriter writer = new FileWriter(writeData);
+            listOfTask.forEach(x-> {
+                try {
+                    writer.write(x.write());
+                    if (x.getIsDone()) {
+                        writer.write("mark " + (listOfTask.indexOf(x) + 1) + "\n");
+                    }
+                } catch (IOException e) {
+                    System.out.println("You do not have access to write to your save file");
+                }
```
This try catch might be a bit redundant since it is enclosed by a try catch that catches the same exception.

howenc replied:
Um, this part here is actually a lamda expression.
So its like a class in itself?
Its like a consumer class?
That means that I need to catch that exception, or at least thats what my IDE said.

howenc replied:
Unless I declare that the class throws a exception?

SelwynAng commented on src/main/java/command/Commands.java:
```diff
@@ -0,0 +1,351 @@
+    private LocalDateTime dateTime;
+
+    /**
+     * Construct a Commands object with only the primary command.
```
Perhaps verbs used in JavaDocs should be singular tense (e.g. Constructs instead of Construct).

howenc replied:
Roger that I'll change

LoMaply commented on src/main/java/task/ListOfTask.java:
```diff
@@ -0,0 +1,171 @@
+        listOfTask.add(temp);
+        if (print) {
+            System.out.println("added: " + temp);
+            Storage.save(listOfTask);
+        }
+    }
+
+    /**
+     * Adds a task into the task list.
+     *
+     * @param task The task that is to be added.
+     * @param dayDate The deadline of the task.
+     * @param print True to print messages, false to not print messages.
+     */
+    public void addTask(String task, LocalDateTime dayDate, boolean print) {
+        Task temp = Task.of(task, dayDate);
+        listOfTask.add(temp);
+        if (print) {
+            System.out.println("added: " + temp);
+            Storage.save(listOfTask);
+        }
+    }
+
+    /**
+     * Adds a task into the task list.
+     *
+     * @param task The task that is to be added.
+     * @param startDayDateTime The date and time of the start of the task.
+     * @param endDayDateTime The date and time of the end of the task.
+     * @param print True to print messages, false to not print messages.
+     */
+    public void addTask(String task, LocalDateTime startDayDateTime,
+                        LocalDateTime endDayDateTime, boolean print) {
+        Task temp = Task.of(task, startDayDateTime, endDayDateTime);
+        listOfTask.add(temp);
+        if (print) {
+            System.out.println("added: " + temp);
+            Storage.save(listOfTask);
+        }
+    }
```
Is there a way to reduce duplication for these 3 addTask methods?

Instead of having each addTask method handle a separate task, you could have a general addTask method taking in a Task object.

This Task object would be created wherever addTask is called as opposed to being created by addTask itself.

howenc replied:
Yup, I made a seperate private method to do this.
I did not want to break encapsulation by allowing a class outside of the task package to touch the more specific tasks.

SelwynAng commented on src/main/java/command/Commands.java:
```diff
@@ -0,0 +1,351 @@
+    }
+
+    private static class TwoCommands extends Commands {
+        private COMMANDS state2;
```
Could there be a better name for this variable? A more relatable variable name would be better!

SelwynAng commented on src/main/java/command/Commands.java:
```diff
@@ -0,0 +1,351 @@
+
+    private static class TwoCommands extends Commands {
+        private COMMANDS state2;
+        private String name2;
```
The same goes for this. A more relatable variable name would be better!

SelwynAng commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,100 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
+public class Storage {
```
Perhaps adding Javadocs for describing this whole class will be better.

SelwynAng commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,179 @@
+        }
+    }
+
+    private static class Deadlines extends Task {
```
Perhaps extracting these private nested classes as standalone public classes would be better?

LoMaply commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,188 @@
+                    Commands c2 = phaseThree.parse();
+                    if (c1.checkCommand(Commands.COMMANDS.FROM) && c2.checkCommand(Commands.COMMANDS.TO)) {
+                        if (c1.compareTime(c2)) {
+                            return Commands.of(Commands.COMMANDS.EVENT, task, c1, c2);
```
Quite deep code indentation here, is there a way to simplify this?

howenc replied:
I'll try, but what do you suggest.

SelwynAng reviewed (commented):
Overall, it was a nice implementation of your Duke project. Keep up the good work!

LoMaply reviewed (commented):
Overall, I like that you took the time to extract both the Exceptions as well as the Commands. Furthermore the testing done using JUnit appears quite in depth.

However the code itself seems quite verbose, to the point where it gets quite hard to follow the logic. There are also several instances of unnecessary duplication.

howenc committed:
Reviewed feedback from PR

howenc committed:
Implemented JavaFx

howenc committed:
Merge branch 'branch-Level-10'

howenc committed:
Implemented CheckStyle

howenc committed:
Minor commit

howenc committed:
Merge branch 'branch-A-CheckStyle'

howenc committed:
Minor edit

howenc committed:
Fixed javaFx and made the GUI

howenc committed:
header  Cleaning up code

description  I have cleaned up some code that is no longer needed as
well as refactoring the execute method of Commands class

howenc committed:
header  Refactoring and Asserting

description  There were some assumptions I made about dukeResponse,
I wanted it to be never be null as this means that duke would reply
with nothing.

Refactoring a few catch blocks according to DRY, removed the need
to use print boolean in delete command.

howenc committed:
Merge pull request #1 from howenc/branch-A-Assertions

header  Refactoring and Asserting

howenc committed:
Merge branch 'master' into branch-A-CodeQuality

howenc committed:
Refactoring and Code Quality

howenc committed:
Merge pull request #2 from howenc/branch-A-CodeQuality

Merge branch 'master' into branch-A-CodeQuality

howenc committed:
header  Implemented undo command

description

- Implemented a stack to function as a log for previous commands
- Implemented a stack to function as a log for deleted tasks

When undoing, it pops the command stack and does the reverse of
its execute method.

howenc committed:
Merge branch 'branch-C-Undo'

howenc committed:
header  Minor edits

howenc committed:
Minor Edits

howenc committed:
Creating a UserGuide

howenc committed:
Improve SLAP

howenc committed:
Update GUI

howenc committed:
Better GUI
